src/main/java/com/researchspace/service/impl/RecordManagerImpl.java

```diff
@@ -1198,20 +1181,28 @@ private List<BaseRecord> getViewableRecordsByRole(User subject, boolean onlyTemp
        *
        */
       List<User> viewableUsers = userDao.getViewableUsersByRole(subject);
-      Set<BaseRecord> recordsToAdd = new HashSet<>();
-      for (User viewableUser : viewableUsers) {
-        recordsToAdd.addAll(getRecordsSharedWithUser(viewableUser, onlyTemplates, subject));
-        recordsToAdd.addAll(getViewableRecordsByUser(viewableUser, onlyTemplates));
-      }
-      result.addAll(recordsToAdd);
-    }
-    // now filter by permission: rspa-2084. We collect into mutable list in case further
-    // modifications are needed to the results.
-    result =
-        result.stream()
-            .filter(br -> permissnUtils.isPermitted(br, PermissionType.READ, subject))
-            .collect(toCollection(() -> new ArrayList<>()));
-    return result;
+      Set<Long> viewableUserIds =
+          viewableUsers.stream().map(User::getId).collect(Collectors.toSet());
+
+      // records shared with the subject
+      Set<BaseRecord> subjectsViewableRecords =
+          new HashSet<>(getRecordsSharedWithUser(subject, onlyTemplates, subject));
+      // add records created by each of the subjects viewable users
+      subjectsViewableRecords.addAll(getViewableRecordsByUsers(viewableUserIds, onlyTemplates));
+
+      // Remove any records the subject is not permitted to see. Using PI role check as a
+      // short-circuit to avoid unnecessary permissions check.
+      // Handles case where the subject is a PI of group A, which contains another PI as a group
+      // member, they can only view the docs of the other PI which has been shared with group A
+      // (whereas the subject PI can view ALL records created by a non-PI), so filter the other PIs
+      // records by their permissions to ensure correct visibility.
+      subjectsViewableRecords.removeIf(
```

Comment:
@rs-nicof now instead of looping through each user, getting their records, then checking permissions if they are a PI, we now get all records in one db call, and have 1 pass through the list, first checking if the user is a PI, then perform a permissions check. 

In most cases the user won't be a PI and the check will short-circuit and avoid checking permissions, and this way avoids multiple db calls to get the initial records.